### PR TITLE
Reorder the BP loop to make manager persistence more reliable

### PR DESCRIPTION
**Pull request by TheBlueMatt**
The main loop of the background processor has this line:
`peer_manager.process_events(); // Note that this may block on ChannelManager's locking`
which does, indeed, sometimes block waiting on the `ChannelManager`
to finish whatever its doing. Specifically, its the only place in
the background processor loop that we block waiting on the
`ChannelManager`, so if the `ChannelManager` is relatively busy, we
may end up being blocked there most of the time.

This should be fine, except today we had a user who's node was
particularly slow in processing some channel updates, resulting in
the background processor being blocked there (as expected). Then,
when the channel updates were completed (and persisted) the next
thing the background processor did was hand the user events to
process, creating yet more channel updates. Ultimately, the users'
node crashed before finishing the event processing. This left us
with an updated monitor on disk and an outdated manager, and they
lost the channel on startup.

Here we simply move the above quoted line to after the normal event
processing, ensuring the next thing we do after blocking on
`ChannelManager` locks is persist the manager, prior to event
handling.

### CODE REVIEW

**Review by vincenzopalazzo — DISMISSED** [review]
LGTM

**Review by arik-so — COMMENTED** [review]
Are we sure this will make a big difference, considering a) it's a loop and b) the other persistence calls are still happening after process_events?

**Comment by TheBlueMatt** [comment]
> Are we sure this will make a big difference

I mean, no, but most "normal" nodes aren't doing a lot of stuff regularly - in the case that led to this PR the node was sending a payment and otherwise not doing anything. In that case, where we're starting from all threads doing nothing, I'd think this will change the "received payment failure case" to retry after persisting the ChannelManager, which is exactly the important change for this case.

> the other persistence calls are still happening after process_events?

which "other persistence calls"? Nothing happens after the event processing now.

**Comment by arik-so** [comment]
I was mostly referring to this line: https://github.com/lightningdevkit/rust-lightning/pull/1436/files#diff-3c70a8bcbb58522dbe589e774a85bbb611adab7142bb1eacec4fb7828cb3ee06R224

But if we're reasonably confident that it was on the first iteration, I'm ACKing

**Review by arik-so — APPROVED** [review]

File: lightning-background-processor/src/lib.rs [hunk 1]
```diff
@@ -203,10 +203,15 @@ impl BackgroundProcessor {
 			let mut have_pruned = false;
 
 			loop {
-				peer_manager.process_events(); // Note that this may block on ChannelManager's locking
 				channel_manager.process_pending_events(&event_handler);
 				chain_monitor.process_pending_events(&event_handler);
 
+				// Note that the PeerManager process_events may block on ChannelManager's locks,
```
Comment (valentinewallace) [inline lightning-background-processor/src/lib.rs:209]: nit: `PeerManager::process_events`

File: lightning-background-processor/src/lib.rs [hunk 2]
```diff
@@ -203,10 +203,15 @@ impl BackgroundProcessor {
...
+				// hence it comes last here. When the ChannelManager finishes whatever its doing,
+				// we want to ensure we get into `persist_manager` as quickly as we can, especially
+				// without running the normal event processing above and handing events to users.
```
Comment (valentinewallace) [inline lightning-background-processor/src/lib.rs:212]: Hmm, not clear to me what "normal event processing" refers to. Also, within `chanman.process_pending_events`, it seems like we *do* hand events to users without calling `persist_manager` first? Missing something..

Comment (TheBlueMatt) [reply lightning-background-processor/src/lib.rs]: I added more docs, let me know if its a bit clearer.

**Review by valentinewallace — APPROVED** [review]

**Comment by TheBlueMatt** [comment]
Squashed fixup.

**Review by valentinewallace — APPROVED** [review]

**Pull request merged**